### PR TITLE
autofix: also run dry-run tests

### PR DESCRIPTION
**Pull request by hannesm**
it unfolds as follows: there was a test_autofix_osemfail and a test_autofix - the former had `@pytest.mark.parametrize("dryrun", [True, False], ids=["dryrun", "not-dryrun"])`, the latter `@pytest.mark.parametrize("dryrun", [False], ids=["not-dryrun"])`

Some time later, we managed to remove the osemfail test definition (since osemgrep passes all these tests) -- and now with this PR, we enable the dry-run tests again (and also make them pass on osemgrep by doing some additional work when converting the matches / when to emit "fixed_lines").

Result for `make osempass` is now 217 passes (and 127 failures) --> 63%

### CODE REVIEW

File: src/osemgrep/reporting/Cli_json_output.ml [hunk 1]
```diff
@@ -345,19 +345,41 @@ let cli_match_of_core_match ~dryrun (hrules : Rule.hrules) (m : OutJ.core_match)
...
-                  let first_line_part = Str.first_chars line (start.col - 1)
-                  and last_line_part =
-                    Str.string_after last_line (end_.col - 1)
-                  in
-                  String.split_on_char '\n'
-                    (first_line_part ^ fix ^ last_line_part)
-              | [], _
-              | _, [] ->
-                  [])
+          Option.fold ~none:None
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.ml]: probably better to move this code in a separate helper function.
Also I'm not a fan of those Option.fold, Option.value ~default ...
in general I find the code that does explciit 'match x with Some y -> ... | None -> ...' clearer that those Option.fold.

Comment (aryx) [reply src/osemgrep/reporting/Cli_json_output.ml]: Option.iter is fine (better than Option.fold ~note:()), 
let* is fine.
But I often find Option.fold and Option.value making the code more complicated.

File: src/osemgrep/reporting/Cli_json_output.mli [hunk 2]
```diff
@@ -14,6 +14,7 @@ val exit_code_of_error_type : Semgrep_output_v1_t.error_type -> Exit_code.t
 (* internals used also for incremental display of matches *)
 val cli_match_of_core_match :
   dryrun:bool ->
+  ?applied_fixes:(string, (int * int) list) Hashtbl.t ->
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.mli:22]: So this hash is actually populated by the function? it's an OUT parameter?
Would it be possible to return it instead with cli_match?
At least add a comment that this parameter is an OUT parameter modifed by the function
and used in the caller to do ... ?

**Review by aryx — APPROVED** [review]

**Review by aryx — CHANGES_REQUESTED** [review]
Request changes for adding comments on the hash OUT parameter and the rewriting of the
helper.

File: src/osemgrep/reporting/Cli_json_output.ml (hunk 1 above)
Comment (hannesm) [reply src/osemgrep/reporting/Cli_json_output.ml]: done in [a943e44](https://github.com/semgrep/semgrep/pull/9648/commits/a943e44754bd4c3cf6bcd5ef96c612d877603677)

File: src/osemgrep/reporting/Cli_json_output.mli (hunk 2 above)
Comment (hannesm) [reply src/osemgrep/reporting/Cli_json_output.mli:22]: I documented the purpose:
```
   The [applied_fixes] hash table is both in and out parameter, it is used for
   deciding whether a fixed_lines elements is included in the cli_match. This
   depends on whether an overlapping fix was already included in an earlier
   cli_match in the same list of matches.
```

hope that makes it clear.

**Review by aryx — APPROVED** [review]

**Pull request merged**